src/spi.rs

```diff
@@ -0,0 +1,314 @@
+//! Serial Peripheral Interface (SPI) module.
+
+use ::nb;
+use ::hal::spi::{FullDuplex, Mode, Phase, Polarity};
+use ::stm32l4x6::{SPI1, SPI2, SPI3};
+
+use ::time::Hertz;
+use ::rcc::{APB1, APB2, Clocks};
+
+use ::ptr;
+
+///Describes GPIO Pins to be used by SPI.
+pub mod gpio {
```

Comment:
The usual approach here (for I2C or UART) is to have a trait:
```rust
pub unsafe trait RxPin<UART> {}
```
and have a pin impl that trait
```rust
unsafe impl RxPin<USART1> for PA1<AF4> {}
```

Why not do that here? I'd imagine a set of traits,
```rust
pub unsafe trait Nss<SPI> {}
pub unsafe trait Sck<SPI> {}
// ...
```

Reply:
I'm not sure what would be the use of these traits though...?

Reply:
I still find this module to be really hard to understand and inflexible. Why not just do as [other crates do](https://github.com/japaric/stm32f30x-hal/blob/master/src/spi.rs#L30) and attach a trait to the GPIO pin structs themselves? It provides all the type safety you need and doesn't require re-exporting existing stuff.

Reply:
There are already traits for each pin, but I just made this module to easily find index of pins per type in this module.
I can remove if you really find it so hard